src/main/java/seedu/address/logic/CommandList.java

```diff
@@ -2,60 +2,86 @@
-    private static final ArrayList<String> inputHistory = new ArrayList<>();
+    private static CommandList theList = null;
+    private static final int MAX_RECORD_NUMBER = 10;
+    private int pointer = -1;
+    private final ArrayList<String> inputHistory = new ArrayList<>();
 
+    private CommandList() {}
     /**
      * Records the latest user input.
      * @param lastCommand the last user input.
      */
-    public static void record(String lastCommand) {
-        if (!lastCommand.equals("history")) {
-            inputHistory.add(lastCommand);
+    public void record(String lastCommand) {
+        if (lastCommand.equalsIgnoreCase(PreviousCommand.COMMAND_WORD)) {
+            return;
         }
+        if (lastCommand.equalsIgnoreCase(NextCommand.COMMAND_WORD)) {
+            return;
+        }
+        inputHistory.add(lastCommand);
         if (inputHistory.size() > MAX_RECORD_NUMBER) {
             clearOldCommand();
+            return;
         }
+        pointer++;
     }
 
     /** Gets the latest user input. */
-    public static String getLastCommand() {
-        if (isEmpty()) {
-            return "";
+    public String getLastCommand() {
+        return inputHistory.get(pointer);
+    }
+
+    /** Checks whether there is any command history or not. */
+    public boolean isEmpty() {
+        return inputHistory.size() == 0;
+    }
+
+    private void clearOldCommand() {
+        inputHistory.remove(0);
+    }
+
+    public static CommandList getList() {
+        if (theList == null) {
+            theList = new CommandList();
         }
-        return inputHistory.get(inputHistory.size() - 1);
+        return theList;
     }
 
-    /** Gets the most recent commands(number capped at MAX_RECORD_NUMBER). */
-    public static String getRecentCommands() {
-        StringBuilder result = new StringBuilder("");
-        for (int i = 1; i <= inputHistory.size(); i++) {
-            result.append(inputHistory.get(inputHistory.size() - i));
-            if (i != inputHistory.size()) {
-                result.append("\n");
-            }
+    /** Increase the pointer value by 1. */
+    public void increasePointer() throws HistoryException {
+        if (pointer >= inputHistory.size() - 1) {
+            throw new HistoryException(NextCommand.MESSAGE_ON_NO_NEXT);
         }
-        return result.toString();
+        pointer++;
     }
 
-    /** Checks whether there is any command history or not. */
-    public static boolean isEmpty() {
-        return inputHistory.size() == 0;
+    /** Decreases the pointer value by 1. */
+    public void decreasePointer() throws HistoryException {
+        if (pointer <= 0) {
+            throw new HistoryException(PreviousCommand.MESSAGE_ON_NO_PREVIOUS);
+        }
+        pointer--;
     }
 
-    private static void clearOldCommand() {
-        inputHistory.remove(0);
+    public void resetPointer() {
+        pointer = inputHistory.size();
```

Comment:
Doesn't this return index out of bounds exception when you try inputHistory.get(pointer) later?

Reply:
no it will not